chia/consensus/blockchain.py

```diff
@@ -917,44 +925,28 @@ async def get_block_generator(
-                previous_block_hash = curr.prev_header_hash
-                prev_block_record = await self.block_store.get_block_record(previous_block_hash)
-                prev_block = await self.block_store.get_full_block(previous_block_hash)
-                assert prev_block is not None
-                assert prev_block_record is not None
-                fork = find_fork_point_in_chain(self, peak, prev_block_record)
-                curr_2: Optional[FullBlock] = prev_block
-                assert curr_2 is not None and isinstance(curr_2, FullBlock)
-                reorg_chain[curr_2.height] = curr_2
-                while curr_2.height > fork and curr_2.height > 0:
-                    curr_2 = await self.block_store.get_full_block(curr_2.prev_header_hash)
-                    assert curr_2 is not None
-                    reorg_chain[curr_2.height] = curr_2
-
             for ref_height in block.transactions_generator_ref_list:
-                if ref_height in reorg_chain:
-                    ref_block = reorg_chain[ref_height]
-                    assert ref_block is not None
-                    if ref_block.transactions_generator is None:
-                        raise ValueError(Err.GENERATOR_REF_HAS_NO_GENERATOR)
-                    result.append(ref_block.transactions_generator)
-                else:
+                if ref_height in additional_height_dict:
                     if ref_height in additional_height_dict:
                         ref_block = additional_height_dict[ref_height]
                         assert ref_block is not None
                         if ref_block.transactions_generator is None:
                             raise ValueError(Err.GENERATOR_REF_HAS_NO_GENERATOR)
                         result.append(ref_block.transactions_generator)
-                    else:
-                        header_hash = self.height_to_hash(ref_height)
-                        if header_hash is None:
-                            raise ValueError(Err.GENERATOR_REF_HAS_NO_GENERATOR)
-                        gen = await self.block_store.get_generator(header_hash)
-                        if gen is None:
-                            raise ValueError(Err.GENERATOR_REF_HAS_NO_GENERATOR)
-                        result.append(gen)
+                elif ref_height in reorg_chain:
+                    ref_block = reorg_chain[ref_height]
+                    assert ref_block is not None
+                    if ref_block.transactions_generator is None:
+                        raise ValueError(Err.GENERATOR_REF_HAS_NO_GENERATOR)
+                    result.append(ref_block.transactions_generator)
+                else:
+                    header_hash = self.height_to_hash(ref_height)
+                    if header_hash is None:
+                        raise ValueError(Err.GENERATOR_REF_HAS_NO_GENERATOR)
+                    gen = await self.block_store.get_generator(header_hash)
```

Comment:
```suggestion
                    [gen] = await self.block_store.get_generators_at([ref_height])
```

Reply:
It could be improved by pulling all generators in one database query, but it wouldn't be so simple to interleave with the generators that don't come from the database